File: .deprecated/Firmware/drivers/keypad.py
"""
Matrix keypad driver for ESP32Calc.
9 rows × 6 cols, columns PULL_UP, rows driven LOW to scan.
"""
from machine import Pin
from time import ticks_ms, sleep_us
from config.pins import ROW_PINS, COL_PINS, KEY_ROWS, KEY_COLS
from config.layout import KEY_MAP
from config.keys import KEY_NONE, KEY_SHIFT, KEY_ALPHA

DEBOUNCE_MS = 12
HOLD_MS     = 600
HOLD_REPEAT_MS = 200


class Keypad:
    def __init__(self, debug=False):
        self._cols = [Pin(p, Pin.IN, Pin.PULL_UP) for p in COL_PINS]
        self._rows = [Pin(p, Pin.OUT, value=1) for p in ROW_PINS]
        self._last_db = [[0] * KEY_COLS for _ in range(KEY_ROWS)]
        self._last_st = [[False] * KEY_COLS for _ in range(KEY_ROWS)]
        self._emit_st = [[False] * KEY_COLS for _ in range(KEY_ROWS)]
        self._alpha = False
        self._shift = False
        self._debug = debug
        if self._debug:
            print("Keypad: rows={} cols={}".format(ROW_PINS, COL_PINS))
            print("Keypad: initial col values={}".format(
                tuple(col.value() for col in self._cols)))

    @property
    def alpha(self):
        return self._alpha

    @property
    def shift(self):
        return self._shift

    def scan(self):
        now = ticks_ms()
        result = KEY_NONE

        for r in range(KEY_ROWS):
            self._rows[r].value(0)
            sleep_us(50)

            for c in range(KEY_COLS):
                k = KEY_MAP[r][c]
                if k == KEY_NONE:
                    continue

                reading = self._cols[c].value() == 0

                if reading != self._last_st[r][c]:
                    self._last_db[r][c] = now

                if now - self._last_db[r][c] >= DEBOUNCE_MS:
                    if reading != self._emit_st[r][c]:
                        self._emit_st[r][c] = reading
                        if reading:
                            if self._debug:
                                print("Keypad: press row={} col={} key={}".format(
                                    r, c, k))
                            if k == KEY_ALPHA:
                                self._alpha = not self._alpha
                                if self._debug:
                                    print("Keypad: alpha={}".format(self._alpha))
                                continue
                            if k == KEY_SHIFT:
                                self._shift = not self._shift
                                if self._debug:
                                    print("Keypad: shift={}".format(self._shift))
                                continue
                            result = k

                self._last_st[r][c] = reading

            self._rows[r].value(1)

        if result != KEY_NONE and self._shift:
            self._shift = False
            if self._debug:
                print("Keypad: shift auto-clear")

        return result

File: .deprecated/.Firmware_deprecated/src/core/keypad.cpp
#include "core/keypad.h"
#include <Arduino.h>
#include <string.h>

static unsigned long last_debounce[KEY_ROWS][KEY_COLS];
static bool last_state[KEY_ROWS][KEY_COLS];
static bool emitted_state[KEY_ROWS][KEY_COLS];

// ── Modifier state ───────────────────────────────────────────────────
static bool alpha_active = false;
static bool shift_active = false;

bool keypad_alpha_active(void) { return alpha_active; }
bool keypad_shift_active(void) { return shift_active; }

// Init matrix GPIOs: columns = INPUT_PULLUP, rows = OUTPUT HIGH
void keypad_init(void) {
    for (int c = 0; c < KEY_COLS; c++) {
        pinMode(COL_PINS[c], INPUT_PULLUP);
    }
    for (int r = 0; r < KEY_ROWS; r++) {
        pinMode(ROW_PINS[r], OUTPUT);
        digitalWrite(ROW_PINS[r], HIGH);
    }
    memset(last_debounce, 0, sizeof(last_debounce));
    memset(last_state, 0, sizeof(last_state));
    memset(emitted_state, 0, sizeof(emitted_state));
    alpha_active = false;
    shift_active = false;
}

// Scan the 9×6 matrix: pull one row LOW at a time, read columns.
// Returns first newly-pressed key after 20ms debounce, or KEY_NONE.
// Modifier keys (ALPHA, SHIFT) are consumed internally — they toggle state
// and return KEY_NONE. Call keypad_alpha_active() / keypad_shift_active()
// to query the current modifier state.
key_code_t keypad_scan(void) {
    unsigned long now = millis();
    key_code_t returned_key = KEY_NONE;

    for (int r = 0; r < KEY_ROWS; r++) {
        digitalWrite(ROW_PINS[r], LOW);
        delayMicroseconds(50);

        for (int c = 0; c < KEY_COLS; c++) {
            key_code_t key = KEY_MAP[r][c];
            if (key == KEY_NONE) continue;

            bool reading = (digitalRead(COL_PINS[c]) == LOW);

            if (reading != last_state[r][c]) {
                last_debounce[r][c] = now;
            }

            if ((now - last_debounce[r][c]) >= DEBOUNCE_MS) {
                if (reading != emitted_state[r][c]) {
                    emitted_state[r][c] = reading;
                    if (reading) {

                        // ── Consume modifier keys ──────────────
                        if (key == KEY_ALPHA) {
                            alpha_active = !alpha_active;
                            continue;          // don't return, don't forward
                        }
                        if (key == KEY_SHIFT) {
                            shift_active = !shift_active;
                            continue;
                        }

                        returned_key = key;
                    }
                }
            }
            
            last_state[r][c] = reading;
        }

        digitalWrite(ROW_PINS[r], HIGH);
    }

    // Auto-release SHIFT after any non-modifier key press
    if (returned_key != KEY_NONE && shift_active) {
        shift_active = false;
    }

    return returned_key;
}
